tls/s2n_client_key_exchange.c

```diff
@@ -49,136 +51,90 @@ static int s2n_rsa_client_key_recv(struct s2n_connection *conn)
-        GUARD(s2n_ecc_params_free(&conn->secure.server_ecc_params));
-    } else {
-        GUARD(s2n_dh_params_free(&conn->secure.server_dh_params));
-    }
-
+    GUARD(s2n_ecc_params_free(&conn->secure.server_ecc_params));
     return 0;
 }
 
 int s2n_client_key_recv(struct s2n_connection *conn)
 {
-    if (conn->secure.cipher_suite->key_exchange_alg->flags & S2N_KEY_EXCHANGE_DH) {
-        return s2n_dhe_client_key_recv(conn);
-    } else {
-        return s2n_rsa_client_key_recv(conn);
-    }
+    const struct s2n_key_exchange_algorithm *kem_core = conn->secure.cipher_suite->key_exchange_alg;
+    struct s2n_blob shared_key = {0};
+
+    GUARD(kem_core->client_key_recv(conn, &shared_key));
+    GUARD(calculate_keys(conn, &shared_key));
+    return 0;
 }
 
-static int s2n_dhe_client_key_send(struct s2n_connection *conn)
+int s2n_dhe_client_key_send(struct s2n_connection *conn, struct s2n_blob *shared_key)
 {
     struct s2n_stuffer *out = &conn->handshake.io;
-    struct s2n_blob shared_key = {0};
+    GUARD(s2n_dh_compute_shared_secret_as_client(&conn->secure.server_dh_params, out, shared_key));
 
-    if (conn->secure.cipher_suite->key_exchange_alg->flags & S2N_KEY_EXCHANGE_ECC) {
-        GUARD(s2n_ecc_compute_shared_secret_as_client(&conn->secure.server_ecc_params, out, &shared_key));
-    } else {
-        GUARD(s2n_dh_compute_shared_secret_as_client(&conn->secure.server_dh_params, out, &shared_key));
-    }
-
-    /* Turn the pre-master secret into a master secret */
-    GUARD(s2n_prf_master_secret(conn, &shared_key));
-
-    /* Erase the pre-master secret */
-    GUARD(s2n_blob_zero(&shared_key));
-    GUARD(s2n_free(&shared_key));
-
-    /* Expand the keys */
-    GUARD(s2n_prf_key_expansion(conn));
+    /* We don't need the server params any more */
+    GUARD(s2n_dh_params_free(&conn->secure.server_dh_params));
+    return 0;
+}
 
-    /* Save the master secret in the cache */
-    if (s2n_allowed_to_cache_connection(conn)) {
-        GUARD(s2n_store_to_cache(conn));
-    }
+int s2n_ecdhe_client_key_send(struct s2n_connection *conn, struct s2n_blob *shared_key)
+{
+    struct s2n_stuffer *out = &conn->handshake.io;
+    GUARD(s2n_ecc_compute_shared_secret_as_client(&conn->secure.server_ecc_params, out, shared_key));
 
     /* We don't need the server params any more */
-    if (conn->secure.cipher_suite->key_exchange_alg->flags & S2N_KEY_EXCHANGE_ECC) {
-        GUARD(s2n_ecc_params_free(&conn->secure.server_ecc_params));
-    } else {
-        GUARD(s2n_dh_params_free(&conn->secure.server_dh_params));
-    }
-
+    GUARD(s2n_ecc_params_free(&conn->secure.server_ecc_params));
     return 0;
 }
 
-static int s2n_rsa_client_key_send(struct s2n_connection *conn)
+int s2n_rsa_client_key_send(struct s2n_connection *conn, struct s2n_blob *shared_key)
 {
     uint8_t client_protocol_version[S2N_TLS_PROTOCOL_VERSION_LEN];
     client_protocol_version[0] = conn->client_protocol_version / 10;
```

Comment:
I know it is not your code... but where does this 10 is coming from? is it a convention? why isn't it a macro? For example GET_VERSION0() and GET_VERSION1()